### PR TITLE
pkg-config: add initial pkg-config files

### PR DESCRIPTION
**Pull request by billatarm**
## Description

Add three package config files for mbedtls, mbedcrypto and mbedx509. Also update various project variables so the generated PC files have the required data needed without hardcoding it everywhere.

This will help distros package the project following existing conventsions between a normal and -devel package that includes the headers and .pc files for pkg-config aware consumers.

Fixes: #228

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided, or not required
- [x] **backport** done, or not required: **Not Required**
- [x] **tests** provided, or not required: **Not Required**

### CODE REVIEW

File: pkgconfig/mbedcrypto.pc.in
Comment (Tachi107) [inline pkgconfig/mbedcrypto.pc.in:4]: These two lines have two issues:

1. `CMAKE_INSTALL_LIBDIR` isn't guaranteed to be a relative path, so you should use a proper path joining function from CMake code and use its result as the value for `libdir`. You can find a full example here: <https://github.com/simdjson/simdjson/pull/1767>
2. Instead of hardcoding `include`, you should use `CMAKE_INSTALL_INCLUDEDIR`, and then do the same as I suggested for `libdir`

The same applies to the other two `.pc.in` files, of course :)

**Review by Tachi107 — COMMENTED** [review]
Hi, thanks for your patch! Distros will be really happy :)

I've left a few suggestions that I think should be addressed before merging this. I'm not a project maintainer, I'm just a guy that played too much with CMake that happens to maintain the mbedtls Debian package.

**Comment by billatarm** [comment]
> Hi, thanks for your patch! Distros will be really happy :)
> 
> I've left a few suggestions that I think should be addressed before merging this. I'm not a project maintainer, I'm just a guy that played too much with CMake that happens to maintain the mbedtls Debian package.

Perfect thanks, i've never touched CMake

File: pkgconfig/mbedcrypto.pc.in
Comment (billatarm) [reply pkgconfig/mbedcrypto.pc.in:4]: @Tachi107 are you proposing for item 2:
```
includedir="${prefix}/@{CMAKE_INSTALL_INCLUDEDIR}"
```
Obviously updated with a path join function.

Comment (billatarm) [reply pkgconfig/mbedcrypto.pc.in]: @Tachi107 suggestion is done, I also added a patch on top to use cmake builtin cmake_path, but I think we would need to bump `cmake_minimum_required(VERSION 3.5.1)` to `cmake_minimum_required(VERSION 3.20)` and I don't know if they will find that tenable.

File: CMakeLists.txt
Comment (billatarm) [reply CMakeLists.txt:45]: That's what I was looking for, wasn't sure where in the build this got set at.

Comment (billatarm) [reply CMakeLists.txt:45]: A lot of things with "VERSION", this script is very subtle on whitespace, i'm surprised they haven't had an issue with it yet.

```bash
[ $VERBOSE ] && echo "Bumping VERSION in CMakeLists.txt"
sed -e "s/ VERSION [0-9.]\{1,\}/ VERSION $VERSION/g" < CMakeLists.txt > tmp
mv tmp CMakeLists.txt
```

Comment (billatarm) [reply CMakeLists.txt:45]: > the project() call you should probably also update that.

Script already handles is:
```bash
./scripts/bump_version.sh --version 3.0.0 --so-crypto 32 --so-x509 39 --so-tls 62
git diff
```
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4a1d04d3b..077894b0f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,7 +42,7 @@ endif()
 project("Mbed TLS" LANGUAGES ${PROJ_LANGS}
     HOMEPAGE_URL "https://www.trustedfirmware.org/projects/mbed-tls/"
     DESCRIPTION "Mbed TLS is a C library that implements cryptographic primitives, X.509 certificate manipulation and the SSL/TLS and DTLS protocols. Its small code footprint makes it suitable for embedded systems."
-    VERSION 3.5.1
+    VERSION 3.0.0
 )
```

**Comment by billatarm** [comment]
> @billatarm is this still draft - or ready for review? thanks

Yes thanks

File: pkgconfig/mbedcrypto.pc.in
Comment (Tachi107) [inline pkgconfig/mbedcrypto.pc.in:2]: Defining `prefix=` is still needed, please have a closer look at the example I gave you at <https://github.com/simdjson/simdjson/pull/1767> ;)

Comment (billatarm) [reply pkgconfig/mbedcrypto.pc.in:2]: > Defining `prefix=` is still needed, please have a closer look at the example I gave you at [simdjson/simdjson#1767](https://github.com/simdjson/simdjson/pull/1767) ;)

Is that actually true?

Does anything actually run `pkg-config --variable=prefix` ? The man [page](https://linux.die.net/man/1/pkg-config) states most packages define it, but doesn't seem to require it. The PC files that iv'e reviewed for popular packages define it as a hardcoded prefix and then reference the variable within the PC file. Its use in that scenario seems justifiable, but with no internal references to prefix within the package file it seems like it can be omitted. Whats the actual use besides someone querying the variable directly vi apkg-config?

Comment (Tachi107) [reply pkgconfig/mbedcrypto.pc.in:2]: It isn't for basic use cases, but having `includedir` reference `prefix` instead of hardcoding the full path makes it possible for users to override the `prefix` variable to make the pkg-config file work even if moved around (i.e. making it somewhat relocatable). It's also a nice convention that's nice to follow, who knows who relies on it for what :)

File: pkgconfig/CMakeLists.txt
Comment (Tachi107) [inline pkgconfig/CMakeLists.txt:3]: `cmake_path(APPEND)` requires CMake 3.20.0, I think it'd be more appropriate for mbedtls to just include the [JoinPaths.cmake](https://github.com/jtojnar/cmake-snips/blob/master/CMakeScripts/JoinPaths.cmake) script. It's very permissively licensed (public domain / MIT)

Comment (billatarm) [reply pkgconfig/CMakeLists.txt:3]: yes that's why I had this is draft to get a comment if they wanted to bump their min version to 3.20  (which is March 2021 release and almost 3 years old) or pull in that script. But since it's a licensing issue, it's their decision. IMO just bump to 3.20.

File: pkgconfig/mbedcrypto.pc.in
Comment (billatarm) [reply pkgconfig/mbedcrypto.pc.in:2]: Yeah but in order to have prefix work as intended I would have to inject `${prefix}` in from the CMake side. The current join script in this patch does not do that.

Comment (billatarm) [reply pkgconfig/mbedcrypto.pc.in:3]: > Yeah but in order to have prefix work as intended I would have to inject `${prefix}` in from the CMake side. The current join script in this patch does not do that.

I dropped the follow up patch, if they don't want to take Jan Tojnar's Cmake script and use the CMake builtin in 3.20, I'll have to figure out how to re-roll that because AFAICT the CMake builtin won't let me pass that through when I quickly tried

File: pkgconfig/mbedcrypto.pc.in
Comment (Tachi107) [inline pkgconfig/mbedcrypto.pc.in:7]: You probably meant [PROJECT_DESCRIPTION](https://cmake.org/cmake/help/latest/variable/PROJECT_DESCRIPTION.html) and [PROJECT_HOMEPAGE_URL](https://cmake.org/cmake/help/latest/variable/PROJECT_HOMEPAGE_URL.html) - the `CMAKE_` variants don't work as expected when mbedtls is used as a subproject.

Comment (williamcroberts) [reply pkgconfig/mbedcrypto.pc.in]: Sorry I have very little experience with cmake, I use autotools mostly. What about the usage of `CMAKE_INSTALL_PREFIX` and in `mbedtls/CMakeLists.txt` the usages of `CMAKE_INSTALL_INCLUDEDIR` and  `CMAKE_INSTALL_LIBDIR`? Also, if it's a subproject, my initial reaction is that it is being built into something else, so are the pc files even relevant?

Comment (Tachi107) [reply pkgconfig/mbedcrypto.pc.in]: > What about the usage of `CMAKE_INSTALL_PREFIX` and in `mbedtls/CMakeLists.txt` the usages of `CMAKE_INSTALL_INCLUDEDIR` and `CMAKE_INSTALL_LIBDIR`?

I only meant the `CMAKE_` variants of `PROJECT_DESCRIPTION` and the other variables generated by the `project()` call, not all variables which start with `CMAKE_` - the others are fine :)

> Also, if it's a subproject, my initial reaction is that it is being built into something else, so are the pc files even relevant?

Yeah, it probably doesn't matter, but there's still no reason to use those top-level variables when you don't need them.

**Comment by billatarm** [comment]
@jtojnar could you comment if we could relicense this Cmake snippet under the dual Apache 2.0/GPL license scheme?

**Comment by jtojnar** [comment]
Sure, feel free to relicense it. The intent is that anyone can use it in any software project.

File: pkgconfig/JoinPaths.cmake
Comment (daverodgman) [inline pkgconfig/JoinPaths.cmake:5]: We very much prefer all contributions to be `Copyright The Mbed TLS Contributors` @jtojnar is this OK with you?

Comment (jtojnar) [reply pkgconfig/JoinPaths.cmake:5]: Sure.

Comment (billatarm) [reply pkgconfig/JoinPaths.cmake]: Done, thanks @jtojnar

**Comment by billatarm** [comment]
> @billatarm We have some CI failure on this PR. Would you be able to check them out please?

The problem seems to be the useages of `HOMEPAGE_URL` and `DESCRIPTION` in the `project` declaration. Those [need](https://cmake.org/cmake/help/latest/command/project.html) 3.12 and 3.9 respectively and those CI builds are on cmake version [3.5.1](https://mbedtls.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-head/detail/PR-8691-head/10/pipeline#step-1615-log-135). Which is also at odds with your own readme requiring `3.10.2` [here](https://github.com/Mbed-TLS/mbedtls/blob/development/README.md?plain=1#L53), but even correcting that would only fix one of the variable usages.

I'll just have to define the variables for now and use those.

**Comment by billatarm** [comment]
> > @billatarm We have some CI failure on this PR. Would you be able to check them out please?
> 
> The problem seems to be the useages of `HOMEPAGE_URL` and `DESCRIPTION` in the `project` declaration. Those [need](https://cmake.org/cmake/help/latest/command/project.html) 3.12 and 3.9 respectively and those CI builds are on cmake version [3.5.1](https://mbedtls.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-head/detail/PR-8691-head/10/pipeline#step-1615-log-135). Which is also at odds with your own readme requiring `3.10.2` [here](https://github.com/Mbed-TLS/mbedtls/blob/development/README.md?plain=1#L53), but even correcting that would only fix one of the variable usages.
> 
> I'll just have to define the variables for now and use those.

Ahh it's documented as why the README and CIs are disparate:
https://github.com/Mbed-TLS/mbedtls/blob/32c28cebb41a25bb4e18d3a530b59c0e5929a324/CMakeLists.txt#L19-L21

**Comment by tom-daubney-arm** [comment]
@billatarm Apologies for the delay on this. I was on leave last week. Thanks for addressing the CI failures. I have kicked off a new CI run just now so we should know if it was successful in a few hours.

**Comment by daverodgman** [comment]
I tweaked the changelog to fix trivial syntax errors, as I think this is the only thing blocking CI - hope that's OK @billatarm

**Comment by williamcroberts** [comment]
> I tweaked the changelog to fix trivial syntax errors, as I think this is the only thing blocking CI - hope that's OK @billatarm

Thanks, appreciate it, yeah `- *`, I wonder why my editor added that, usually retext is pretty good.

**Review by tom-daubney-arm — APPROVED** [review]
I think this all looks sensible. I have been playing around with it and it seems to work as I would expect, and the CMake stuff all looks sensible. Thanks @billatarm. LGTM.

**Comment by williamcroberts** [comment]
> The version field is empty in all pkgconfig files. Is that expected?

No, and I can reproduce, not sure what happened.

**Comment by billatarm** [comment]
> The version field is empty in all pkgconfig files. Is that expected?

No and I can reproduce, let me see whats up

**Comment by billatarm** [comment]
> > The version field is empty in all pkgconfig files. Is that expected?
> 
> No and I can reproduce, let me see whats up

Ok, so in the final version of the patch when I removed the HOMEPAGE_URL and DESCRIPTION, and let that go back to original form, it broke, I dunno why, and github doesn't show my the version delats AFAIK. But this patch fixes it:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5585c78fa..3d6f14f4b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -33,10 +33,17 @@ cmake_policy(SET CMP0011 NEW)
 # is deprecated and will be removed in future versions.
 cmake_policy(SET CMP0012 NEW)
 
+
 if(TEST_CPP)
-    project("Mbed TLS" LANGUAGES C CXX)
+    project("Mbed TLS"
+           LANGUAGES C CXX
+           VERSION 3.5.2
+    )
 else()
-    project("Mbed TLS" LANGUAGES C)
+    project("Mbed TLS"
+           LANGUAGES C
+           VERSION 3.5.2
+    )
 endif()
 
 include(GNUInstallDirs)
```

I'll submit a PR

File: ChangeLog.d/pkg-config-files-addition.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/pkg-config-files-addition.txt:1]: We're announcing a new feature, but this isn't tested. Are you planning a follow-up with a test (either expanding an existing cmake-related component or a new component in `tests/scripts/all.sh`)?

Comment (billatarm) [reply ChangeLog.d/pkg-config-files-addition.txt:1]: > We're announcing a new feature, but this isn't tested. Are you planning a follow-up with a test (either expanding an existing cmake-related component or a new component in `tests/scripts/all.sh`)?

Yeah the goal is to add this to automated tests doing a make install and then running `pkg-config`, I have to look at how the test suite is constructed to see where the best place is.

Comment (billatarm) [reply ChangeLog.d/pkg-config-files-addition.txt:1]: See PR for tests: https://github.com/Mbed-TLS/mbedtls/pull/8988

File: pkgconfig/mbedtls.pc.in
Comment (Tachi107) [inline pkgconfig/mbedtls.pc.in:9]: Late feedback: shouldn't this be `Requires` instead of `Requires.private`? The CMake code, for instance, has a `PUBLIC` dependency on `mbedx509`.

https://github.com/Mbed-TLS/mbedtls/blob/30978ec6507c0a0376ae041fa333a2f24b0c09b8/library/CMakeLists.txt#L314-L320

Comment (Biswa96) [reply pkgconfig/mbedtls.pc.in:9]: That depends on the following question. If a program links with mbedtls _only_ dynamically, does it need to link with mbedcrypto and mbedx509 directly?

Comment (billatarm) [reply pkgconfig/mbedtls.pc.in:9]: Requires.private is correct here. Those libraries are linked to the shared object not the requestor.

I think this explains it better than me:
- https://stackoverflow.com/questions/45441250/how-does-the-linking-process-differ-when-using-requires-vs-requires-private-in-p

Comment (Tachi107) [reply pkgconfig/mbedtls.pc.in:9]: > That depends on the following question. If a program links with mbedtls _only_ dynamically, does it need to link with mbedcrypto and mbedx509 directly?

I was under the impression that this was the case, but no, `Requires.private` seems appropriate. `mbedtls/ssl.h` includes `mbedtls/x509_crt.h`, which is part of `mbedx509`, so the user ends up needing the appropriate _compiler_ flags for `mbedx509` too, but unless it uses x509 functions it does not need its _linking_ flags. Due to [multiple reasons](https://gitlab.freedesktop.org/pkg-config/pkg-config/-/issues/7), pkg-config's `Requires.private` propagates the `Cflags` when doing shared builds, but not the `Libs`, so this is exactly what we need.

In short: the pkg-config files are ok, but `mebdtls`' `PUBLIC` dependency on `mbedx509` in CMake is probably wrong.

**Pull request merged**
